core/src/main/java/io/grpc/CallOptions.java

```diff
@@ -103,10 +118,12 @@ public CallOptions withCompression(@Nullable String compressorName) {
    * @param deadlineNanoTime the deadline in the clock as per {@link System#nanoTime()}.
    *                         {@code null} for unsetting the deadline.
    */
+  @Deprecated
   public CallOptions withDeadlineNanoTime(@Nullable Long deadlineNanoTime) {
-    CallOptions newOptions = new CallOptions(this);
-    newOptions.deadlineNanoTime = deadlineNanoTime;
-    return newOptions;
+    Deadline deadline = deadlineNanoTime != null
+        ? Deadline.after(deadlineNanoTime - System.nanoTime(), TimeUnit.NANOSECONDS)
```

Comment:
@ejona86 but `getDeadlineNanoTime()` now returns `System.nanoTime()` if the deadline is expired, but it would return some value less than `System.nanoTime()` before.

Depending on how people use these methods (e.g. to compare values), it might break their code i.e. one would expect the below to always be true
```java
opts.withDeadlineNanoTime(nanos).getDeadlineNanoTime() == nanos
```

We could retain the original behaviour by making the `Deadline` constructor package private and introduce a package private method `Deadline.deadlineNanos()`, which we immediately deprecate as well. WDYT?